src/Stratis.Bitcoin/P2P/SelfEndpointTracker.cs

```diff
@@ -0,0 +1,25 @@
+﻿using System.Net;
+using ConcurrentCollections;
+
+namespace Stratis.Bitcoin.P2P
+{
+    /// <summary>
+    /// Tracker for endpoints known to be self.
+    /// </summary>
+    public class SelfEndpointTracker : ISelfEndpointTracker
+    {
+        /// <summary>Hashset to hold the endpoints currently known to be itself.</summary>
+        private readonly ConcurrentHashSet<IPEndPoint> knownSelfEndpoints = new ConcurrentHashSet<IPEndPoint>();
+
+        /// <inheritdoc/>
+        public void Add(IPEndPoint ipEndPoint)
+        {
+            this.knownSelfEndpoints.Add(ipEndPoint);
```

Comment:
Dont you need a TryAdd here?

Reply:
ConcurrentHashSet only uses Try for remove, Add will just add and if it is already there it doesn't matter, it can't really fail.

Reply:
Super ok let push